examples/single_inverter_voltage_current_control_safe_opt.py

```diff
@@ -0,0 +1,243 @@
+
+        return -error.squeeze()
+
+
+if __name__ == '__main__':
+    #####################################
+    # Definitions for the GP
+    prior_mean = 2  # mean factor of the GP prior mean which is multiplied with the first performance of the initial set
+    noise_var = 0.001 ** 2  # measurement noise sigma_omega
+    prior_var = 2  # prior variance of the GP
+
+    # Choose Kp and Ki (current and voltage controller) as mutable parameters (below) and define bounds and lengthscale
+    # for both of them
+    bounds = [(0.0, 0.03), (0, 300),(0.0, 0.03), (0, 300)] # bounds on the input variable current-Ki&Kp and voltage-Ki&Kp
+    lengthscale = [.005, 50.,.005, 50.]   # length scale for the parameter variation [current-Ki&Kp and voltage-Ki&Kp] for the GP
+
+    # The performance should not drop below the safe threshold, which is defined by the factor safe_threshold times
+    # the initial performance: safe_threshold = 1.2 means: performance measurement for optimization are seen as
+    # unsafe, if the new measured performance drops below 20 % of the initial performance of the initial safe (!)
+    # parameter set
+    safe_threshold = 2
+
+    # The algorithm will not try to expand any points that are below this threshold. This makes the algorithm stop
+    # expanding points eventually.
+    # The following variable is multiplied with the first performance of the initial set by the factor below:
+    explore_threshold = 2
+
+    # Factor to multiply with the initial reward to give back an abort_reward-times higher negative reward in case of
+    # limit exceeded
+    abort_reward = 10
+
+    # Definition of the kernel
+    kernel = GPy.kern.Matern32(input_dim=len(bounds), variance=prior_var, lengthscale=lengthscale, ARD=True)
+
+    #####################################
+    # Definition of the controllers
+    # Choose Kp and Ki for the current and voltage controller as mutable parameters
+    mutable_params = dict(currentP=MutableFloat(10e-3), currentI=MutableFloat(10), voltageP=MutableFloat(25e-3),
+                            voltageI=MutableFloat(60))
+
+    voltage_dqp_iparams = PI_params(kP=mutable_params['voltageP'], kI=mutable_params['voltageI'], limits=(-iLimit, iLimit))
+    current_dqp_iparams = PI_params(kP=mutable_params['currentP'], kI=mutable_params['currentI'], limits=(-1, 1))
+
+    # Define the droop parameters for the inverter of the active power Watt/Hz (DroopGain), delta_t (0.005) used for the
+    # filter and the nominal frequency
+    # Droop controller used to calculate the virtual frequency drop due to load changes
+    droop_param = DroopParams(DroopGain, 0.005, nomFreq)
+
+    # Define the Q-droop parameters for the inverter of the reactive power VAR/Volt, delta_t (0.002) used for the
+    # filter and the nominal voltage
+    qdroop_param = DroopParams(QDroopGain, 0.002, nomVoltPeak)
+
+    # Define a voltage forming inverter using the pi pi and droop parameters from above
```

Comment:
....PIPI control and drop parameters....

Reply:
done